internal/argocd/revision.go

```diff
@@ -10,68 +11,163 @@ import (
 	kargoapi "github.com/akuity/kargo/api/v1alpha1"
 	argocd "github.com/akuity/kargo/internal/controller/argocd/api/v1alpha1"
 	"github.com/akuity/kargo/internal/controller/freight"
+	"github.com/akuity/kargo/internal/logging"
 )
 
 // GetDesiredRevision returns the desired revision for the given
 // v1alpha1.Application. If that cannot be determined, an empty string is
 // returned.
-func GetDesiredRevision(
+func GetDesiredRevisions(
 	ctx context.Context,
 	cl client.Client,
 	stage *kargoapi.Stage,
 	update *kargoapi.ArgoCDAppUpdate,
 	app *argocd.Application,
 	frght []kargoapi.FreightReference,
-) (string, error) {
+) ([]string, error) {
+
+	logger := logging.LoggerFromContext(ctx)
+	if app == nil {
+		err := errors.New("Application is nil")
+		logger.Error(err, "Cannot determine desired revision for application, bailing.")
+		return nil, nil
+	}
+
+	appLogger := logger.WithValues("application", app.Name, "namespace", app.Namespace)
+	appLogger.Debug("Getting desired revision for application", "app", app, "spec", app.Spec)
+
 	// Note that frght was provided as an argument instead of being plucked
 	// directly from stage.Status, because this gives us the flexibility to use
 	// this function for finding the revision to sync to either in the context of
 	// a health check (current freight) or in the context of a promotion (new
 	// freight).
-	switch {
-	case app == nil || app.Spec.Source == nil:
-		// Without an Application, we can't determine the desired revision.
-		return "", nil
-	case app.Spec.Source.Chart != "":
-		// This source points to a Helm chart.
-		// NB: This has to go first, as the repository URL can also point to
-		//     a Helm repository.
+
+	// An application may have one or more sources.
+	if !app.IsMultisource() {
+		s := app.Spec.Source
+		if s == nil {
```

Comment:
spec.Source and spec.Sources are mutually exclusive, right?

I think this logic can be simplified...

```
if app.Spec.Source != nil {
  // Zeroes in on single source app and doesn't require a second conditional
  // ...
  // return ...
}
// Everything else false through to here

Reply:
Actually, considering this further, I don't think you need to special-case single source.

```
sources := app.Spec.Sources
if app.Spec.Source != nil {
  sources = []argocd.ApplicationSource{*app.Spec.Source}
}
// Now loop over all sources; no need for a special case
```